transport/tlsfrag/buffer.go

```diff
@@ -81,28 +77,32 @@ func (b *clientHelloBuffer) Write(p []byte) (int, error) {
 // If this buffer still requires more data to build a complete TLS Client Hello message, it returns nil error.
 //
 // You can call ReadFrom multiple times if r doesn't provide enough data to build a complete Client Hello packet.
+//
+// ReadFrom will hang indefinitely if r provides fewer than 5 bytes and doesn't return the io.EOF error (e.g., "PING").
 func (b *clientHelloBuffer) ReadFrom(r io.Reader) (n int64, err error) {
 	// Waiting to finish the header of 5 bytes
-	for len(b.data) < recordHeaderLen {
-		m, e := r.Read(b.data[len(b.data):recordHeaderLen])
+	if len(b.data) < recordHeaderLen {
+		m, e := io.ReadFull(r, b.data[len(b.data):recordHeaderLen])
 		b.data = b.data[:len(b.data)+m]
 		n += int64(m)
-
-		if len(b.data) == recordHeaderLen {
-			hdr := tlsRecordHeader(b.data)
-			if err = validateTLSClientHello(hdr); err != nil {
-				b.validationErr = err
-				return
-			}
-			buf := make([]byte, 0, recordHeaderLen*2+hdr.PayloadLen())
-			b.data = append(buf, b.data...)
-		}
 		if err = e; err != nil {
-			if err == io.EOF {
+			if err == io.EOF || err == io.ErrUnexpectedEOF {
 				err = nil
```

Comment:
The non-nil `err` here means `r` did not provide enough bytes to construct the header (5 bytes), it either `EOF`'ed or encountered other errors. It's expected to return `err` or `nil` (representing `EOF`) here because that's what the original `Reader` `r` returns.

And we need to wait for more bytes to construct the full header, the user can either call `Write`, or `ReadFrom` another `Reader`, later.

I've added the `*SingleByteSeq` test cases to illustrate this behavior: each time `r` will only produce one byte and then `EOF`. This is a valid scenario for example when the caller wants to send requests from multiple files, and the caller will call `ReadFrom` against each 1-byte file.